core/include/detray/tools/navigator.hpp

```diff
@@ -260,27 +254,71 @@ class navigator {
     DETRAY_HOST_DEVICE
     navigator(const detector_t &d) : _detector(&d) {}
 
+    template <typename track_t>
+    DETRAY_HOST_DEVICE inline intersection update_next_candidate(
+        state &navigation, const track_t &track) const {
+
+        dindex obj_idx = navigation.next()->index;
+        auto sfi =
+            intersect(track, _detector->surfaces()[obj_idx],
+                      _detector->transform_store(), _detector->mask_store());
+        sfi.index = obj_idx;
+        (*navigation.next()) = sfi;
+        navigation.set_dist(sfi.path);
+
+        return sfi;
+    }
+
     /** Navigation status() call which established the current navigation
      *  information.
      *
-     * @tparam track_t type of the track (including context)
+     * @tparam stepper_state_t type of stepper state
      *
      * @param navigation [in, out] is the navigation state object
-     * @param track [in] is the track infromation
+     * @param stepping the stepper state
      *
      * @return a heartbeat to indicate if the navigation is still alive
      **/
-    template <typename track_t>
+    template <typename stepper_state_t>
     DETRAY_HOST_DEVICE inline bool status(state &navigation,
-                                          const track_t &track) const {
-
+                                          stepper_state_t &stepping) const {
         bool heartbeat = true;
 
-        // If there is no_trust (e.g. at the beginning of the navigation in a
-        // volume), the kernel will be initialized. Otherwise the candidates
-        // are re-evaluated based on current trust level
-        update_kernel(navigation, track,
-                      _detector->volumes()[navigation.volume()]);
+        const auto &track = stepping();
+
+        if (navigation.trust_level() >= e_high_trust) {
+            while (not navigation.is_exhausted()) {
+
+                // Only update the next candidate
+                auto sfi = update_next_candidate(navigation, track);
+
+                if (sfi.status == e_inside) {
```

Comment:
Should this not also be conatin  ```release_step_size```?

Reply:
`target` contains it so it doesn't have to be there anymore. Actually that's why I restructured the navigator

Reply:
What does ```status()``` do now, if some actor in between decided we fall below high trust (I know we currently only have the stepper and it is unlikely to do that)? What if this hypothetical actor changes the track state somehow and we don't update the step size accordingly? I think its safer and easier to understand if we communicate all distance changes to the stepper

Reply:
I am sorry for my ignorance on Actor - I don't understand very well...   Maybe you can show some simple example codes 

> if some actor in between decided we fall below high trust 

you mean `fair_trust`? Yeah I removed it because current navigator doesn't have any use case of it. but we can take it onto the surface once we need it.

> we don't update the step size accordingly?

Are you sure? step size is released in `target` right before the `step` call.

Reply:
I am not sure what the actors might or might not do either. Andi seemed to think there could be a usecase for these trust levels, so I preserved them until I could see the full picture with more complex actors in detray.

Reply:
In that case ```status()``` sets the trust level to ```full_trust``` and the target call should need to do nothing

Reply:
No, I meant the ```status()``` call does nothing right now, if we are below high trust, or am I getting this wrong? 
```if (navigation.trust_level() >= e_high_trust) {```

Reply:
I can put the `fair_trust` case back into status or target

Reply:
Maybe I will put it in `target` cause it seems more reasonable to me

Reply:
OK